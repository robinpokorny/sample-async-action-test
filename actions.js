export const FETCH_TODOS_ERROR = 'FETCH_TODOS_REJECTED'
export const FETCH_TODOS_START = 'FETCH_TODOS_PENDING'
export const FETCH_TODOS_SUCCESS = 'FETCH_TODOS_FULFILLED'

export const loadTodos = () => ({ getState, fetch }) => {
  const { todos } = getState()
  const url = todos.get('serverUrl')

  return ({
    type: 'FETCH_TODOS',
    payload: fetch(url, {
      credentials: 'same-origin',
      headers: {
        'Content-Type': 'application/json'
      }
    })
  })
}
