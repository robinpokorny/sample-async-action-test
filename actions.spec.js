/* eslint-env jest */
import { isFSA } from 'flux-standard-action'
import { fromJS } from 'immutable'

import * as actions from './actions'

describe('Todo action', () => {
  describe('loadTodos', () => {
    const action = actions.loadTodos()

    const state = {
      todos: fromJS({
        serverUrl: 'XURL'
      })
    }
    const deps = {
      fetch: jest.fn(() => Promise.resolve()),
      getState: () => state
    }

    beforeEach(() => {
      deps.fetch.mockClear()
    })

    it('is a valid FSA', () => {
      expect(isFSA(action(deps))).toBe(true)
    })

    it('has correct data', () => {
      expect(action(deps)).toMatchSnapshot()
    })

    it('fetches todos from the server', () => {
      action(deps)

      expect(deps.fetch).toHaveBeenCalledTimes(1)
      expect(deps.fetch.mock.calls[0]).toMatchSnapshot()
    })
  })
})
