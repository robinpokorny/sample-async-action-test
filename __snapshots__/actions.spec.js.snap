exports[`Todo action loadTodos has correct data 1`] = `
Object {
  "payload": Promise {},
  "type": "FETCH_TODOS",
}
`;
